### PR TITLE
jira - module revamp -> moved code to class

### PR DESCRIPTION
**Pull request by russoz**
##### SUMMARY
Major revamp in the module, keeping the same behaviour. Code should be simpler to maintain in the future.

(Following up on https://github.com/ansible-collections/community.general/pull/1978#issuecomment-813924732)

Wait for #2192 to merge - after that this PR will require a rebase.

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
jira

### CODE REVIEW

**Comment by felixfontein** [comment]
Let's keep it [WIP] until #2192 is merged.

File: plugins/modules/web_infrastructure/jira.py
Comment (russoz) [inline plugins/modules/web_infrastructure/jira.py:577]: Hi @DWSR , Pycham is moaning about this line:
- apparently `base64.decode()` returns `None` so the content used in line 585 below is null
- `base64.decode()` expects two parameters

Could you please double check on this?

Comment (russoz) [reply plugins/modules/web_infrastructure/jira.py:577]: Are you just testing if the content *is* in base64 or do you really need to decode it? I would expect the former, but asking to be on the safe side. If indeed it's the first case, then maybe what you need is:
```suggestion
                dummy = base64.b64decode(content)
```

Comment (DWSR) [reply plugins/modules/web_infrastructure/jira.py:577]: It needs to be decoded. The `content` param for the module is expected to be a base64 string in order to handle situations where the file content itself is binary (which Ansible can't handle).

Not sure how I missed this either. 🤦

File: plugins/modules/web_infrastructure/jira.py
Comment (russoz) [inline plugins/modules/web_infrastructure/jira.py:620]: This is the bug reported in #2234 . I have added a few integration tests - certainly not enough - but we should also get unit tests going to ensure small things like this don't make it through the cracks.

**Comment by aminvakil** [comment]
@russoz Could you please rebase your branch?

**Comment by russoz** [comment]
@aminvakil I will hold this back for a while - this is in dire need of proper unit tests, so I will tend to them before asking for review again. Thanks

**Comment by russoz** [comment]
Hi @aminvakil @felixfontein @DWSR I still want to add unit tests, but I'm not with a lot of time this week, probably not in the next either. So I'd rather push this one in sooner rather than later, will add the tests later on. 

Could you please review it? TIA

File: plugins/modules/web_infrastructure/jira.py
Comment (felixfontein) [reply plugins/modules/web_infrastructure/jira.py]: It would be great if this could be moved to a separate PR so it can be backported to stable-2.

File: plugins/modules/web_infrastructure/jira.py
Comment (felixfontein) [inline plugins/modules/web_infrastructure/jira.py:467]: Is this correct? This function did not returned `True` for changed in the old version.

File: plugins/modules/web_infrastructure/jira.py
Comment (felixfontein) [inline plugins/modules/web_infrastructure/jira.py:472]: Is this correct? This function did not returned `True` for changed in the old version.

File: plugins/modules/web_infrastructure/jira.py
Comment (felixfontein) [reply plugins/modules/web_infrastructure/jira.py]: Or, well, there should probably be a separate PR on stable-2 fixing that for stable-2 once this one is merged which is mostly unrelated to this PR, except parts of the changelog fragment appear in it and this change.

File: plugins/modules/web_infrastructure/jira.py
Comment (felixfontein) [inline plugins/modules/web_infrastructure/jira.py:602]: As a later refactoring, it probably makes sense to remove `user` and `passwd`, since they are always the same.

Comment (russoz) [reply plugins/modules/web_infrastructure/jira.py:602]: Yup, it makes sense, removing those and `timeout` as well

**Comment by russoz** [comment]
the base64 decode bug has moved to #2349

**Comment by felixfontein** [comment]
@russoz can you please rebase?

File: plugins/modules/web_infrastructure/jira.py
Comment (felixfontein) [inline plugins/modules/web_infrastructure/jira.py:525]:
```suggestion
    def operation_attach(self):
```
Otherwise this won't work I think.

**Review by felixfontein — APPROVED** [review]

**Comment by felixfontein** [comment]
@russoz thanks for improving this module!
@DWSR thanks for reviewing!

**Pull request merged**
